downloader.go

```diff
@@ -240,41 +239,51 @@ func (d *Downloader) prepareAndStartDownload(ctx context.Context, url string, wg
 		ch <- s
 		return
 	}
-	defer f.Close()
 	if err := f.Truncate(int64(t)); err != nil {
 		s.Error = fmt.Errorf("error truncating %s to %d: %w", path, t, err)
 		ch <- s
+		wg.Done()
 		return
 	}
+	var urlDownload sync.WaitGroup
 	for _, c := range d.chunks(t) {
-		b, err := d.downloadChunk(ctx, url, c)
-		if err != nil {
-			s.Error = err
-			ch <- s
-			return
-		}
-		_, err = f.WriteAt(b, int64(c.start))
-		if err != nil {
-			s.Error = fmt.Errorf("error writing to %s: %w", path, err)
+		urlDownload.Add(1)
+		go func(c chunk) {
+			defer urlDownload.Done()
+			b, err := d.downloadChunk(ctx, url, c)
+			if err != nil {
+				s.Error = err
+				ch <- s
+				return
+			}
+			_, err = f.WriteAt(b, int64(c.start))
+			if err != nil {
+				s.Error = fmt.Errorf("error writing to %s: %w", path, err)
+				ch <- s
+				return
+			}
+			s.DownloadedFileBytes += int64(len(b))
 			ch <- s
-			return
-		}
-		s.DownloadedFileBytes += int64(len(b))
-		ch <- s
+		}(c)
 	}
+	go func() {
+		urlDownload.Wait()
+		wg.Done()
+		f.Close()
+	}()
 }
 
 // DownloadWithContext is a version of Download that takes a context. The
 // context can be used to stop all downloads in progress.
 func (d *Downloader) DownloadWithContext(ctx context.Context, urls ...string) <-chan DownloadStatus {
 	if d.client == nil {
-		d.client = &http.Client{Timeout: d.TimeoutPerChunk}
+		d.client = newClient(d.MaxParallelDownloadsPerServer, d.TimeoutPerChunk)
 	}
-	ch := make(chan DownloadStatus)
-	var wg sync.WaitGroup
+	ch := make(chan DownloadStatus, 2*len(urls)) // the first status will be the total file size (and or an error creating/trucating the file).
+	var wg sync.WaitGroup                        // this wait group is used to wait for all chunks (from all downloads) to finish.
 	for _, u := range urls {
 		wg.Add(1)
-		go d.prepareAndStartDownload(ctx, u, &wg, ch)
+		d.prepareAndStartDownload(ctx, u, &wg, ch)
```

Comment:
NIT Maybe too much of a detail, but we can avoid dealing with two wait groups inside `prepareAndStartDownload` if we wrap this as:

```go
wg.Add(1)
go func(u string) {
    d.prepareAndStartDownload(ctx, u, ch)
    wg.Done()
}(u)
```

Reply:
For that to be correct, we must synchronously inside prepAndStartDownload.